File: api/src/main/java/dev/obscuria/fragmentum/registry/DeferredItem.java
package dev.obscuria.fragmentum.registry;

import net.minecraft.core.Holder;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.ItemLike;
import org.apache.commons.lang3.NotImplementedException;

import java.util.function.Supplier;

@SuppressWarnings("unused")
public class DeferredItem<T extends Item> extends Deferred<Item, T> implements ItemLike
{
    public DeferredItem(Supplier<Holder<Item>> supplier)
    {
        super(supplier);
    }

    @Override
    public Item asItem()
    {
        throw new NotImplementedException();
    }

    public ItemStack instantiate()
    {
        throw new NotImplementedException();
    }
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/registry/DeferredItem.kt
@file:Suppress("unused", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.registry

import net.minecraft.core.Holder
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.ItemLike

class DeferredItem<T : Item>(supplier: () -> Holder<Item>) : Deferred<Item, T>(supplier), ItemLike {

    override fun asItem(): Item = value

    fun instantiate(): ItemStack = asItem().defaultInstance
}